### ISSUE TITLE
I wonder what is the shape of y_repd in AC loss?

### ISSUE THREAD

**Issue by haozhi1817** [xuuuuuuchen/Active-Contour-Loss#6]
Hi @xuuuuuuchen 

In another issue, you said that the input shape in AC loss is [Batch, 1, W, H].

Dose  it mean that the y_pred is an output of softmax layer followed by an argmax operator or a pick operator to pick out the probability of fg(or bg)?

If the shape of y_pred is [Batch, 1, W, H], why do you pick the first channel (y_pred[:, 0, :, :]) when you implement the Eq[12] but not do so when you implement the Eq[11] ?

**Comment by xuuuuuuchen** [comment]
Hi @haozhi1817 ,
1. y_pred is an output of the sigmoid layer because it is a 2-class segmentation problem.
2. It depends on how you pre-define the shape C_1/_2. If the shape of C_1/_2 was firstly defined as np.ones/zeros((256, 256)), y_pred must have a same size like [:, 0, :, :]. You could use y_pred directly without slicing out the 1st channel, if so, C_1/_2 need to be pre-defined as np.ones/zeros((**1**, 256, 256)).

**Comment by qaqzzz** [comment]
class ActiveContourLoss(Module):
    def __init__(self):
        super(ActiveContourLoss, self).__init__()
    
    def forward(self, y_pred, y_true, combine=None):
    
        x = y_pred[:,:,1:,:] - y_pred[:,:,:-1,:] # horizontal and vertical directions 
        y = y_pred[:,:,:,1:] - y_pred[:,:,:,:-1]
        
        delta_x = x[:,:,1:,:-2]**2
        delta_y = y[:,:,:-2,1:]**2
        delta_u = torch.abs(delta_x + delta_y)
        
        epsilon = 0.00000001 # where is a parameter to avoid square root is zero in practice.
        w = 1.
        
        if combine is not None:
            lenth = w * torch.mean(torch.sqrt(delta_u + epsilon)) # equ.(11) in the paper
        else:
            lenth = w * torch.sum(torch.sqrt(delta_u + epsilon)) # equ.(11) in the paper
    
        if torch.cuda.is_available():
            C_1 = torch.ones(y_true.shape, dtype=torch.float32).cuda()
            C_2 = torch.zeros(y_true.shape, dtype=torch.float32).cuda()
        else:
            C_1 = torch.ones(y_true.shape, dtype=torch.float32)
            C_2 = torch.zeros(y_true.shape, dtype=torch.float32)
        
        if combine is not None:
            region_in = torch.abs(torch.mean(y_pred * ((y_true - C_1)**2)) ) # equ.(12) in the paper
            region_out = torch.abs(torch.mean((1. - y_pred) * ((y_true - C_2)**2))) # equ.(12) in the paper
        else:
            region_in = torch.abs(torch.sum(y_pred * ((y_true - C_1)**2)) ) # equ.(12) in the paper
            region_out = torch.abs(torch.sum((1. - y_pred) * ((y_true - C_2)**2))) # equ.(12) in the paper
            
        lambdaP = 5. # lambda parameter could be various.
        loss = lenth + lambdaP * (region_in + region_out) 

        return loss
